Update extension to use RMP's new graphQL API

MoralCode opened the pull request:
oh boy this includes a lot of things so uuh i apologize in advance if you want to code review this line by line...

also dont look too close at the history unless you want your brain to hurt - i changed my mind on things like the build process a number of times while getting it to work haha

Changes made:

- setup webpack build system to automatically bundle dependencies together (should be documented in README)
- restructured/organized the code in the repo into a `src` file tree
- create some classes to hold the data for a professor and the ratings (this helps separate the data from the presentation and make it easier to swap out if the API ever changes later)
- use the new graphQL api instead of the old rest one
    - this removes a lot of logic for things like retrying different names 

Testing:
Tested on firefox
Tested on opera (essentially chrome) 99

Changes to be made:

- [x] version number bump
- [x] maybe add handling for some float values that have rounding errors
- [x] improved handling for "no data" scenarios like removing "N/A/5" and "-1%"
- [x] filtering to avoid queries of "To be determined" professors
- [x] set extension ID in the manifest for firefox
- [ ] release to the various extension stores (after more complete testing)
    - [X] figure out if #3 is a problem for deploying to the stores and resolve it (may require splitting the build process back out into two folders like initially had it before combining it 



Screenshots:
![Screenshot_20230622_114912](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/17362949/468d3743-9ab3-4b15-80c0-2cffa41c9f48)
![Screenshot_20230622_120505](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/17362949/9ec5a8fb-8953-4a41-8915-0b328b175713)

fixes #2 
fixes #3

MoralCode committed:
install graphQL package

MoralCode committed:
add new function to get ratings

MoralCode committed:
rearrange new function gql query to include parameters

MoralCode committed:
use the new function instead

MoralCode committed:
update endpoint in manifest

MoralCode committed:
firefox specific manifest updates

MoralCode committed:
add more parameters to the graphQL api

MoralCode committed:
move files into a src tree

MoralCode committed:
manifest tweaks

MoralCode committed:
npm update

MoralCode committed:
first pass

MoralCode committed:
convert ico into pngs with imagemagick
convert "favicon.ico" -thumbnail 16x16 -alpha on -background none -flatten "favicon.png"

MoralCode committed:
gitignore dist folder

MoralCode committed:
fix manifest lint errors

MoralCode committed:
add lint script to package.json

MoralCode committed:
eslint setup

MoralCode committed:
re-add content script

MoralCode committed:
add new schedulemaker url

MoralCode committed:
update RMP url

MoralCode committed:
remove webpack

MoralCode committed:
add build script from https://github.com/tosdr/browser-extensions/commit/b82d413a0cd4b9174ea8b2096cc7d403d17669c1

MoralCode committed:
move firefox specific manifest things to a separate manifest that gets combined later

MoralCode committed:
use airbnb base linter

MoralCode committed:
ignore build dir

MoralCode committed:
copy lixin.js from https://github.com/tosdr/browser-extensions/commit/c14702eef36d72758745ee1169c19bf09212b4e7

MoralCode committed:
first pass manifest generation

MoralCode committed:
fix manifest generation

MoralCode committed:
add dev more to build.sh that doesnt delete the build dir

MoralCode committed:
dont try and lint tabs to spaces

MoralCode committed:
document build process

MoralCode committed:
build script edits

MoralCode committed:
fix indentation linter

MoralCode committed:
run a lint fix pass

MoralCode committed:
add lint fix command

MoralCode committed:
lint fixes to webpack config

MoralCode committed:
add no-tabs rule

MoralCode committed:
update build process again

MoralCode committed:
remove firefox specific build step - manifestv3 requires it for chrome too

MoralCode committed:
unify the two extensions (remove separate build directories)

MoralCode committed:
install web-ext

MoralCode committed:
images are no longer copied to src

MoralCode committed:
rename files into new tree

MoralCode committed:
linter run

MoralCode committed:
Merge branch 'webpack' into graphql

MoralCode committed:
reorder imports

MoralCode committed:
remove platform dirs in webpack config

MoralCode committed:
eslint is annoying

MoralCode committed:
convert webpack thing into an .mjs

MoralCode committed:
prevent accidental publishing

MoralCode committed:
fix missing EJS __dirname thanks to https://stackoverflow.com/a/72462507/

MoralCode committed:
fix package.json import
https://stackoverflow.com/a/66276798

MoralCode committed:
setup webpack bundles
https://webpack.js.org/guides/output-management/

MoralCode committed:
remove file manager plugin

MoralCode committed:
install web-ext plugin

MoralCode committed:
remove build.sh from build scripts

MoralCode committed:
remove more eslint junk

MoralCode committed:
tehre arent really stages anymore - just webpack

MoralCode committed:
rename dev and prod scripts

MoralCode committed:
Merge branch 'webpack' into graphql

MoralCode committed:
lockfile got rearranged

MoralCode committed:
fix missing papa import

MoralCode committed:
background wrapper is unused

MoralCode committed:
fix web request permissions

MoralCode committed:
export helpers so they actually get bundled

MoralCode committed:
use browser.runtime (maybe that helps?)

MoralCode committed:
make a custom message handler to send different types of queries

MoralCode committed:
import arrive so the content runs

MoralCode committed:
pass graphql queries to the background worker to make

MoralCode committed:
tweaks to graphQL query to make it acceptable to RMP's API

MoralCode committed:
apparrently dont need the graphql library (its a server lib)

MoralCode committed:
return graphQL json to frontend

MoralCode committed:
receive response from background

MoralCode committed:
add a function to normalize some of the graphQL data

MoralCode committed:
add an RMPData class

MoralCode committed:
start populating RMPData class

MoralCode committed:
rename to RMPProfessorData

MoralCode committed:
finish constructor for RMPProfessorData

MoralCode committed:
add RMPRatingData class

MoralCode committed:
add WouldTakeAgainCount to professor data

MoralCode committed:
parse professor data from graphQL

MoralCode committed:
add IDs to professor data class

MoralCode committed:
tweak graphQL constructor for professor data based on graphQL structure

MoralCode committed:
update RMP Rating data based on GraphQL structure

MoralCode committed:
translate data from GraphQL into an RMPProfessorData

MoralCode committed:
add some extra graphQL fields to the query

MoralCode committed:
set up some scaffolding for displaying a single rating

MoralCode committed:
add some fields to the rating data part of the query

MoralCode committed:
GetProfessorRatingNew doesn't require an element

MoralCode committed:
create a crude copy of the tooltip setup code - modified to use prodFata

MoralCode committed:
tweak a couple data defaults for professorData so presence can easily be checked

MoralCode committed:
add some professor helpers to generate certain data more conveniently

MoralCode committed:
remove code iterating through all ratings - we have that

MoralCode committed:
configure devtool so that both development and prod mode have source maps that dont use eval()

MoralCode committed:
make data constructors static so they can be used as intended

MoralCode committed:
cleanup more old code and link up the new tooltip function

MoralCode committed:
import tippy

MoralCode committed:
fix a couple bugs that were causing the data to not be complete

MoralCode committed:
remove some old print lines in the background script

MoralCode committed:
include content.css in the final builds

MoralCode committed:
popper is already a dependency of tippy.js

MoralCode committed:
include css for the popup

MoralCode committed:
make tooltip target source more clear than "this"

MoralCode committed:
use cleaner way to give html content to tippy

MoralCode committed:
remove allowHTML tag to prevent XSS - it seems to work without it anyway sicne we arent adding HTML by string

MoralCode committed:
remove npm test script - its now built into the build one

MoralCode committed:
remove old RMP API code

MoralCode committed:
update README

MoralCode committed:
allow builds to overwrite the last built output file (removes scary red error)

CalvinWu4 commented:
after running `npm run build-dev or npm run build`
I get this when I try to drag the  .zip file into chrome://extensions/
![image](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/3fcecaa5-614a-45d0-b81c-792df3797c30)

MoralCode commented:
Ah yep, i know what that is. Thats a remnant of me testing on FF only and not chrome.

If you want to manually/temporarily fix so it only works on chrome, change the `background` key in the manifest to be the following object:

```
{
 "service_worker": "<the same filepath to the background script>"
}
```

Will fix more permanently soon

MoralCode committed:
getNameCombos is now unused

MoralCode committed:
install polyfill for all browsers

MoralCode committed:
set up config as a function accepting env for customization
https://stackoverflow.com/a/44114071/

MoralCode committed:
reformat webExtPlugin settings

MoralCode committed:
apply a tweak to the manifest for firefox extensions
thanks mozilla https://github.com/mozilla/web-ext/issues/2532

MoralCode committed:
update README with new build commands

MoralCode committed:
add DEV flag to dev build zips

MoralCode committed:
add a release script shortcut

MoralCode committed:
typo

MoralCode committed:
add cs to web accessible resources to appease chrome

MoralCode committed:
"Google Chrome emits an "Invalid match pattern" error if the pattern has a path other than '/*'"
https://developer.chrome.com/docs/extensions/mv3/manifest/web_accessible_resources/

MoralCode committed:
fix firefox dev script to actually make a firefox build

MoralCode committed:
nother build script typo

MoralCode committed:
add some bundles and things to the web accessible resources to appease chrome

MoralCode committed:
tell chrome to use a dynamic URL

MoralCode commented:
ok, NOW it should work in chrome.

why cant the two browsers agree on friggin standards already in the year 2023

MoralCode commented:
big questions:
- what version number should this be released as? 3.0? 2.0.2? 2.1?
- whats this extensions firefox addon ID? Ive been using a dummy value in the manifest so far

CalvinWu4 commented:
3.0
<img width="398" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/998d240b-1bd2-4011-86a2-a14b2c72a359">

MoralCode committed:
add real firefox ID to manifest

MoralCode committed:
version bump

MoralCode commented:
awesome, thanks!
Do you have any thoughts on the three other items in the "Changes to be made" section of the PR description? they dont seem super critical but may be nice to have

let me know if theres anything else you want me to add as well!

CalvinWu4 commented:
<img width="272" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/88708ecb-ff61-4457-8f14-154a9572c4d0">

hmm im seeing the link to rmp link but not the rating or the tooltip
EDIT: o nvm you don't add it if they are not found

CalvinWu4 commented:
and yeah it would be nice to handle the float values

CalvinWu4 commented:
![image](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/1b385e63-db9e-4c50-8288-7190caba7261)
![image](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/9f274de8-52e1-4041-b844-889e8abaef7f)

also i think you put the level of difficulty rating as the overall quality rating?

MoralCode committed:
set up unit testing

MoralCode committed:
merge addedNicknames.js into utils

MoralCode committed:
fix firefox id

MoralCode committed:
move constants to the top of utils

MoralCode committed:
update import in content script

MoralCode committed:
use .mjs extensions now that the whole package is a module

MoralCode committed:
add a tested function for filtering professor names

MoralCode committed:
Merge branch 'test' into graphql

MoralCode committed:
apply filtering to remove "to be determined" names professors

MoralCode committed:
fix difficulty and quality ratings

MoralCode committed:
round data as it comes in

MoralCode committed:
document existence of tests

MoralCode committed:
add fallback if wouldTakeAgainPercentage is less than 0

MoralCode committed:
caching attempt

MoralCode committed:
save data to cache

MoralCode committed:
Revert "caching attempt"

all the requests are happening concurrently so its not effective

MoralCode commented:
> and yeah it would be nice to handle the float values

done


> also i think you put the level of difficulty rating as the overall quality rating?

ah whoops, fixed this too


Did you have any kind of rate limiting in place before? or do you pretty much just spam the API? 
Do you think it would be wise to set the `User-Agent` to point to this GH repo so that if someone at RMP gets mad they can file an issue?

Other Additions:
- added a little filtering for names that are not professors (like the TBD ones from scheduleMaker)
- tried to add some in memory storage of retrieved professor data to try and prevent duplicate requests for the same exact professor when they appear in a list multiple times, but it didnt work

MoralCode committed:
split the linking of professor data into its own function in case thats useful later

MoralCode committed:
bump version number

MoralCode committed:
update web-ext plugin

MoralCode committed:
remove linter bodge job and replace it with an actual update to web-ext plugin
this adds a feature that suppresses certain errors that happen on known-good chrome extensions

CalvinWu4 commented:
![image](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/462698ea-4cbb-4157-8d89-ad396d606199)
getting this when i try to build chrome now

MoralCode commented:
have you done an npm install since pulling the most recent few commits? theres a contribution i made upstream to the web-ext plugin that filters out this error (i.e. prevents the lint from stopping the whole build).

This linter error is known to happen on chrome extensions that chrome says is fine because the linter was written by firefox and the two are disagreeing on how to do background script stuff in manifest v3....

CalvinWu4 commented on src/content/index.js:
```diff
@@ -0,0 +1,268 @@
+		thisEasyRating.classList.add('prof-rating-text');
+		thisEasyRating.textContent = `Level of Difficulty: ${mostHelpfulReview.difficultyRating}`;
+		div.appendChild(thisEasyRating);
+		if (mostHelpfulReview.rWouldTakeAgain !== 'N/A') {
+			const thisWouldTakeAgain = document.createElement('div');
+			thisWouldTakeAgain.classList.add('prof-rating-text');
+			thisWouldTakeAgain.textContent = `Would take again: ${mostHelpfulReview.iWouldTakeAgain}`;
+			div.appendChild(thisWouldTakeAgain);
+		}
```
Could you fix that so that it doesn't show up as undefined?

MoralCode replied:
ah, looks like i forgot to double-check a few things relating to the more helpful review.

CalvinWu4 commented on src/content/index.js:
```diff
@@ -0,0 +1,268 @@
+		div.appendChild(classText);
+		const dateText = document.createElement('div');
+		dateText.classList.add('prof-rating-text');
+		dateText.textContent = mostHelpfulReview.date; //TODO: maybe convert to string date
```
`dateText.textContent = new Date(mostHelpfulReview.date).toLocaleDateString();`

MoralCode replied:
done!

MoralCode committed:
refector the creation of the WouldTakeAgain part of the popup so that it also works within the "most helpful review" portion

MoralCode committed:
fill in some missing review data

MoralCode committed:
refactor tag creation part to handle the case where there are no tags for a rating

MoralCode committed:
refactor again to create a generally-reusable element creation function

MoralCode committed:
remove specific wouldTakeAgainElement creation function in favor of more general one

MoralCode committed:
convert the timestamps of reviews into date objects

MoralCode committed:
convert review date to a locale date string per code review

MoralCode committed:
refactor the creation of elements to massively clean up the tooltip generation code

MoralCode committed:
some random autoformatting changes

MoralCode commented:
fixed the two review comments and also refactored the process for building the elements of the tooltip so that it used a common function

CalvinWu4 commented:
> have you done an npm install since pulling the most recent few commits? theres a contribution i made upstream to the web-ext plugin that filters out this error (i.e. prevents the lint from stopping the whole build).
> 
> This linter error is known to happen on chrome extensions that chrome says is fine because the linter was written by firefox and the two are disagreeing on how to do background script stuff in manifest v3....

it doesnt fix it
this commit is causing it
https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/commit/649d79df6245242150e7effd3f097f0e37badc6e

CalvinWu4 commented:
im not seeing the tooltip work anymore with the new commits

MoralCode committed:
fix error with string processing causing things to crash

MoralCode commented:
should work now - was a small error from the refactor/cleanup that was causing things to crash

CalvinWu4 committed:
Revert "remove linter bodge job and replace it with an actual update to web-ext plugin"

This reverts commit 649d79df6245242150e7effd3f097f0e37badc6e.

CalvinWu4 committed:
fix prof rating showing difficulty rating vs quality rating

CalvinWu4 commented:
not sure where to add this in the code
```function addLinkForProfNotFound(element, profData) {
	const schoolId = 'U2Nob29sLTgwNw==';
	element.setAttribute('target', '_blank');
	element.classList.add('blueText');
	element.parentElement && element.parentElement.classList.add('classSearchBasicResultsText');
	element.setAttribute('href', `https://www.ratemyprofessors.com/search/teachers?query=${profData.lastName}&sid=${schoolId}`);
```

CalvinWu4 commented on contentscript.js:
```diff
@@ -1,302 +0,0 @@
-                GetProfessorRating(element, fullName, lastName, originalLastName, firstName, originalFirstName, 
-                    middleNames, originalMiddleNames, tryNicknames, nicknamesIndex, tryMiddleAndLastNameCombos,
-                    middleAndLastNameCombosIndex, tryMiddleNameAsFirst, tryMiddleNames);
-            }
-            // Try again with nicknames for the first name
-            else if (tryNicknames && nicknames[originalFirstName]) {
-                firstName = nicknames[originalFirstName][nicknamesIndex];
-                nicknamesIndex++;
-                tryNicknames = nicknames[originalFirstName][nicknamesIndex];
-                if (!tryNicknames) {
-                    restoreFirstName = true;
-                }
-                GetProfessorRating(element, fullName, lastName, originalLastName, firstName, originalFirstName, 
-                    middleNames, originalMiddleNames, tryNicknames, nicknamesIndex, tryMiddleAndLastNameCombos,
-                    middleAndLastNameCombosIndex, tryMiddleNameAsFirst, tryMiddleNames);
-            }
-            // Try again with the middle name as the first name
-            else if (tryMiddleNameAsFirst && !tryMiddleNames && originalMiddleNames.length > 0) {
-                tryMiddleNameAsFirst = false;
-                restoreFirstName = true;
-                GetProfessorRating(element, fullName, lastName, originalLastName, firstName, originalFirstName, 
-                    middleNames, originalMiddleNames, tryNicknames, nicknamesIndex, tryMiddleAndLastNameCombos,
-                    middleAndLastNameCombosIndex, tryMiddleNameAsFirst, tryMiddleNames);
-            }
-            // Try again with middle names
-            else if (!tryMiddleNames && originalMiddleNames.length > 0) {
-                tryMiddleNames = true;
-                GetProfessorRating(element, fullName, lastName, originalLastName, firstName, originalFirstName, 
-                    middleNames, originalMiddleNames, tryNicknames, nicknamesIndex, tryMiddleAndLastNameCombos,
-                    middleAndLastNameCombosIndex, tryMiddleNameAsFirst, tryMiddleNames);
-            }
-            // Set link to search results if not found
-            else {
-                element.textContent = `${element.textContent} (NF)`;
-                element.setAttribute('href', 
-                `https://www.ratemyprofessors.com/search/teachers?query=${originalLastName}&sid=${schoolId}`);
-            }
-        }        
-    });
-}
```
can we restore the logic i had here for searching names?

MoralCode replied:
I could, but you might have to help me understand what logic you were going for as I had a hard time trying to untangle it from the old RMP API queries and ended up just starting over from scratch.

What I'm kinda going for with the new code is like a two-stage approach of:
1. Generate a list of search strings
2. Try them out one by one until you get results

Do you think you could do the first step of looking at the old code and refactoring it into a function that takes in something representing a professors name and generates a list of search strings to try in order of likelihood?

I can provide basically any value that exists in the first part of the `arrive` callback (i.e. raw name from the page, filtered name, name as a list of strings per part, etc) and i also have a list of nearly every professors name at RIT that i scraped from ScheduleMaker if you want that for testing.

This restructure should make it easier to change the api code in future if needed without affecting the name searching process

CalvinWu4 replied:
Can we simplify the logic to this?

Given a [First Name] [First Middle Name?] [Second Middle Name?] [Last Name]

try
[First Name] [Last Name]
[First Middle Name (that is not a surname particle)] [Last Name]
[Second Middle Name (that is not a surname particle)] [Last Name]
[try all nicknames for First Name] [Last Name]
[try all nicknames for First Middle Name (that is not a surname particle)] [Last Name]
[try all nicknames for Second Middle Name (that is not a surname particle)] [Last Name]
if no luck, then provide a link with just [Last Name]

MoralCode replied:
Oh ok, that makes sense, ill see if i can write that up into code when i get a chance

MoralCode commented:
~~Why did you make commit d2317e25961889eeb7fcc560b40e1c618f461eb5 to revert the fix i made to the linter configuration? Was it not working for you? Should I have done a better job at documenting that this fix requires updating to the latest version of the linting plugin in order to work?~~

> it doesnt fix it
> this commit is causing it
> [649d79d](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/commit/649d79df6245242150e7effd3f097f0e37badc6e)

what error are you seeing? what about it is not working? you might need to run `npm update` insteadl of `npm install` but as long as you have `web-ext-plugin` version `2.8.0` installed it should work.

Edit: i see that plugin may be causing a crash - will work on that

MoralCode commented:
> not sure where to add this in the code

This is already in the code at the top of the `displaySingleProfRating` function. The RateMyProfessors site now uses a new URL scheme that looks like `https://www.ratemyprofessors.com/professor/1809609` using a professors `legacyId` instead of their name and school ID.

MoralCode committed:
Revert "Revert "remove linter bodge job and replace it with an actual update to web-ext plugin""

This reverts commit d2317e25961889eeb7fcc560b40e1c618f461eb5.

CalvinWu4 commented:
> > not sure where to add this in the code
> 
> This is already in the code at the top of the `displaySingleProfRating` function. The RateMyProfessors site now uses a new URL scheme that looks like `https://www.ratemyprofessors.com/professor/1809609` using a professors `legacyId` instead of their name and school ID.

<img width="1437" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/a2d75445-9b64-4761-aa62-758c1a83ef44">
im not seeing the link when a professor is not found

MoralCode commented:
Ohhhhh in that case you may have had the correct URL format for performing a search. My apologies, I didnt realize that the extension linked every professors name regardless of whether they were found or not.


Should there be special handling if theres a preexisting rate my professors link on the professors name? (As is the case on schedulemaker). Do you want to assume that each individual website (SM, TC, etc) thats generating its own RMP links knows how to do it better (for the profs at that institution) than a general purpose implementation (and thus skip replacing the links) or should this extension overwrite any links that may be present?

MoralCode committed:
remove New from GetProfessorRating

MoralCode committed:
refactor check for surname particle into a function

MoralCode committed:
create a utility function to generate name combinations to search for

MoralCode committed:
split out the name searching process into a function

MoralCode committed:
finish the search term generation util

MoralCode committed:
some simple unit tests and associated fixes for the search term generation

MoralCode committed:
update readme to reflect new location of the added nicknames list

MoralCode committed:
allow school ID to be passed in for a particular RMP query

MoralCode committed:
make searchProfessorByName async and give it a maxTries variable

MoralCode committed:
update linkProfessor to handle both the success and failure case

MoralCode committed:
create a function to releatedly search successive names in the search list asyncronously

MoralCode committed:
pass in a school name and perform the search based on the strings

MoralCode committed:
supply nicknames to the search generation process

MoralCode committed:
fail early if there is no professor name to search for

MoralCode committed:
make RIT school ID a constant and pass it in when searching

MoralCode committed:
link the professor name regardless of success or failure

CalvinWu4 commented:
> Ohhhhh in that case you may have had the correct URL format for performing a search. My apologies, I didnt realize that the extension linked every professors name regardless of whether they were found or not.
> 
> Should there be special handling if theres a preexisting rate my professors link on the professors name? (As is the case on schedulemaker). Do you want to assume that each individual website (SM, TC, etc) thats generating its own RMP links knows how to do it better (for the profs at that institution) than a general purpose implementation (and thus skip replacing the links) or should this extension overwrite any links that may be present?

Yeah you can keep the ScheduleMaker links

MoralCode committed:
check for and avoid messing with existing rateMyProfessors links

MoralCode committed:
bump to web-ext plugin 2.8.1

MoralCode commented:
Tested again in chrome and firefox - both seem to be working as expected in TC and ScheduleMaker - any other fixes that need to be made that you can see?

MoralCode committed:
tweaks to dev and prod build environments

CalvinWu4 commented:
<img width="414" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/595b8f95-d2bf-43be-a886-a9c73cd5931d">
<img width="191" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/ecf4a57f-3971-4b2a-908b-fefff076528f">

why am i seeing a discrepancy btwn the values here?

MoralCode commented:
```bash
curl 'https://www.ratemyprofessors.com/graphql' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/114.0' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br' -H 'Referer: https://www.ratemyprofessors.com/professor/568275' -H 'Content-Type: application/json' -H 'Authorization: Basic dGVzdDp0ZXN0' -H 'Origin: https://www.ratemyprofessors.com' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Cookie: ccpa-notice-viewed-02=true; cid=-VEdhkoW5x-20230620; userSchoolId=U2Nob29sLTgwNw==; userSchoolLegacyId=807; userSchoolName=Rochester%20Institute%20of%20Technology' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Sec-GPC: 1' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache' --data-raw '{"query":"query NewSearchTeachersQuery( $query: TeacherSearchQuery!) { newSearch { teachers(query: $query) { didFallback edges { cursor node { id legacyId firstName lastName avgRatingRounded numRatings wouldTakeAgainPercentRounded wouldTakeAgainCount teacherRatingTags { id legacyId tagCount tagName } mostUsefulRating { id class isForOnlineClass legacyId comment helpfulRatingRounded ratingTags grade date iWouldTakeAgain qualityRating difficultyRatingRounded teacherNote{ id comment createdAt class } thumbsDownTotal thumbsUpTotal } avgDifficultyRounded school { name id } department } } } } }","variables":{"query":{"text":"Samuel Malachowsky","schoolID":"U2Nob29sLTgwNw=="}}}'
```

This is basically the query that the extension uses to fetch the professor (i just put the name in manually). run it yourself to see what data the extension gets.
The values look close enough that it appears to be a rounding problem on RMP's side. I suspect RMP's official web frontend will round something like `"wouldTakeAgainPercentRounded": 93.10000000000001` to 94%, whereas this extension correctly handles it and displays 93.1%. The same appears to be true for `"avgDifficultyRounded": 2.8000000000000003,`. `"avgRatingRounded": 4.6000000000000005,` may have some algorithm rounding it up to the next interval of 0.2, but it still seems loosely consistent with the theory.

CalvinWu4 commented:
Ok this looks really wrong though
<img width="427" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/52258605-f143-473b-8acc-4214d83701c6">

For this [prof](https://www.ratemyprofessors.com/professor/2539311#ratingsList), I manually calculated the rating to be 2.33 but we have 1.7 somehow?

MoralCode commented:
if you exclude the 5.0 and a 4.0 (i.e. the two newest datapoints) the average drops to 1.7. The api data we get says `numRatings` is 7 whereas the site shows 9 so that also provides some kind of supporting evidence. Evidence for what though, I don't know. Some theories:
- Maybe we are intentionally getting limited/older data to prevent scraping (this api access seems quite generous).
- Maybe its a "free tier API" thing? (idk if they even have a paid tier tho).
     - Maybe its because we are using test credentials in the auth header (same as their own site is)
- Maybe their graphql api is just a cache meant to speed up searching or something? The page loads for a specific prof dont seem to make any graphql queries so presumably their backend that generates those professor pages has more direct access to read the latest thing from the database.
- Maybe they have some kind of filtering in place to prevent bots/brigading/other forms of manipulating professors scores

The extension is faithfully showing the data that RMP gives us back in response to our queries. Beyond that almost anything could be happening on RMP's side that could cause this. Reaching out to them might be better than my wild guessing.

CalvinWu4 committed:
Update README.md

CalvinWu4 commented on src/manifest.json:
```diff
@@ -0,0 +1,69 @@
+		"https://ratemyprofessors.com/graphql",
+		"https://www.ratemyprofessors.com/graphql",
+		"https://raw.githubusercontent.com/carltonnorthern/nickname-and-diminutive-names-lookup/master/names.csv/",
+		"webRequest"
```
![image](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/ef234278-f1aa-41dd-8aba-0490797e2900)
this is causing this on the Chrome Web Store
is it safe to delete this?

MoralCode replied:
that may be another chrome vs firefox difference, if possible i would keep it in for firefox. 

there should be code that modifies the base manifest for whichever version is not the default - i believe this is at the top of the webpack config (you should see it manipulating the background/service worker entries to make the browsers happy. that would be the place to do it

CalvinWu4 replied:
yeah it works w/o the webrequest in chrome

CalvinWu4 commented:
also are you seeing this for the firefox build?
<img width="674" alt="image" src="https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/26044296/51f0a56b-68c0-407f-bcd9-4fb9eb11d10d">

MoralCode commented:
![Screen Shot 2023-07-23 at 10 06 53](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/17362949/a6aa0438-002a-443f-a6ee-efb4d2529197)

here is a screenshot of what i see while performing the same search and cropping to about the same place.

i suspect you are seeing all "not founds" because of permissions (you might be able to see the errors if you open inspect element and/or the inspection tool for the background worker and click the search button again).

if you find the addon and use the menu to "manage extension"  you should be able to see a permissions tab - double check that all the boxes are checked. I notice that these sometimes get reset, especially the first time you load the extension manually from the debugging menu . I suspect that these permissions will be requested from users at install time in the production build, but I have not tested this.

![Screenshot_20230723_101000](https://github.com/CalvinWu4/RIT-Rate-My-Professors-Extension/assets/17362949/7b058b2c-f281-458b-a439-7a3b5b865f63)

CalvinWu4 commented:
> ![Screen Shot 2023-07-23 at 10 06 53](https://user-images.githubusercontent.com/17362949/255398272-a6aa0438-002a-443f-a6ee-efb4d2529197.png)
> 
> here is a screenshot of what i see while performing the same search and cropping to about the same place.
> 
> i suspect you are seeing all "not founds" because of permissions (you might be able to see the errors if you open inspect element and/or the inspection tool for the background worker and click the search button again).
> 
> if you find the addon and use the menu to "manage extension" you should be able to see a permissions tab - double check that all the boxes are checked. I notice that these sometimes get reset, especially the first time you load the extension manually from the debugging menu . I suspect that these permissions will be requested from users at install time in the production build, but I have not tested this.
> 
> ![Screenshot_20230723_101000](https://user-images.githubusercontent.com/17362949/255398309-7b058b2c-f281-458b-a439-7a3b5b865f63.png)

Ok that fixes it
Adding the add-on to Firefox from here: https://addons.mozilla.org/en-US/firefox/addon/rate-my-professors-for-rit/ doesn't prompt for those permissions though
I think that can be fixed by editing the manifest.json?

MoralCode commented:
so when you install the extension from prod in a fresh environment and it will show all NF's because it needs perms it didnt ask for?

CalvinWu4 commented:
> so when you install the extension from prod in a fresh environment and it will show all NF's because it needs perms it didnt ask for?

yes

MoralCode commented:
and this is only on firefox? not chrome?

CalvinWu4 commented:
> and this is only on firefox? not chrome?

yes

MoralCode commented:
seems like host_permissions are treated as optional by most browsers.
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/host_permissions

Tried a few ideas for how to fix this and haven't found one that works well yet

CalvinWu4 commented:
I've pretty sure all we have to do is add those sites into host permissions

MoralCode commented:
> I've pretty sure all we have to do is add those sites into host permissions

which sites?

CalvinWu4 commented:
Nvm we need to add the code here
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/request

MoralCode commented:
oh wow, sorry i forgot about this PR. Did you find out whether the permission request process works on firefox without the added code you mentioned?

CalvinWu4 commented:
that still remains to be done
